### PR TITLE
fix: Implement gRPC errors with proper format #44

### PR DESCRIPTION
**Pull request by adbusa67**
#### Description:
This pull request adds gRPC status codes in errors and adds error source as constants. for issue #44 

#### Present Behaviour:
octerr implemented but does not align with gRPC errors.

#### Expected Behaviour:
use status gRPC util to implement errors.

### CODE REVIEW

File: internal/controller/server/repository/job/job_test.go [hunk 1]
```diff
@@ -76,7 +78,7 @@ func TestCheckJobIsAvailableForJobNotAvailable(t *testing.T) {
 	mockClient.On("GetValue", "metadata/testJobName").Return("", errors.New(constant.NoValueFound))
 
 	jobAvailabilityStatus, err := jobRepository.CheckJobIsAvailable(context.Background(), "testJobName")
-	assert.Equal(t, "job with testJobName name not found", err.Error())
+	assert.Equal(t, status.Error(codes.NotFound, errors.New(constant.Etcd+"job with testJobName name not found").Error()).Error(), err.Error())
```
Comment (JaiminRathod) [inline internal/controller/server/repository/job/job_test.go:81]: change errors.New(constant.Etcd+"job with testJobName name not found").Error() to constant.Etcd+"job with testJobName name not found"

File: internal/controller/server/repository/job/job_test.go [hunk 2]
```diff
@@ -88,7 +90,7 @@ func TestCheckJobIsAvailableForEtcdClientFailure(t *testing.T) {
 	mockClient.On("GetValue", "metadata/testJobName").Return("", errors.New("failed to get value from database"))
 
 	jobAvailabilityStatus, err := jobRepository.CheckJobIsAvailable(context.Background(), "testJobName")
-	assert.Equal(t, "failed to get value from database", err.Error())
+	assert.Equal(t, status.Error(codes.Internal, errors.New("failed to get value from database").Error()).Error(), err.Error())
```
Comment (JaiminRathod) [inline internal/controller/server/repository/job/job_test.go:93]: change errors.New("failed to get value from database").Error() to "failed to get value from database"

File: internal/controller/server/repository/job/job_test.go [hunk 3]
```diff
@@ -110,7 +112,7 @@ func TestDeleteForEtcdClientFailure(t *testing.T) {
 
 	mockClient.On("DeleteKey").Return(false, errors.New("failed to delete key from database"))
 	err := jobRepository.Delete(context.Background(), "12345")
-	assert.Equal(t, "failed to delete key from database", err.Error())
+	assert.Equal(t, status.Error(codes.Internal, errors.New("failed to delete key from database").Error()).Error(), err.Error())
```
Comment (JaiminRathod) [inline internal/controller/server/repository/job/job_test.go:115]: Change as suggested above

File: internal/controller/server/repository/job/job_test.go [hunk 4]
```diff
@@ -170,7 +172,7 @@ func TestFetchNextJobForEtcdClientFailure(t *testing.T) {
 
 	assert.Nil(t, nextExecutionData)
 	assert.Equal(t, "", nextJobID)
-	assert.Equal(t, err.Error(), "failed to get keys and values from database")
+	assert.Equal(t, err.Error(), status.Error(codes.Internal, errors.New("failed to get keys and values from database").Error()).Error())
```
Comment (JaiminRathod) [inline internal/controller/server/repository/job/job_test.go:175]: Change as suggested above

File: internal/controller/server/repository/job/job_test.go [hunk 5]
```diff
@@ -188,7 +190,7 @@ func TestFetchNextJobForJobNotAvailable(t *testing.T) {
 
 	assert.Nil(t, nextExecutionData)
 	assert.Equal(t, "", nextJobID)
-	assert.Equal(t, err.Error(), "no pending job in pending job list")
+	assert.Equal(t, err.Error(), status.Error(codes.NotFound, errors.New(constant.Controller+"no pending job in pending job list").Error()).Error())
```
Comment (JaiminRathod) [inline internal/controller/server/repository/job/job_test.go:193]: Change as suggested above

File: internal/controller/server/repository/metadata/metadata.go [hunk 6]
```diff
@@ -36,25 +38,25 @@ func (c *metadataRepository) Save(ctx context.Context, key string, metadata *cli
...
 	if err != nil {
-		return nil, err
+		return nil, status.Error(codes.Internal, err.Error())
 	}
 	dbKey := constant.MetadataPrefix + key
 
 	gr, err := c.etcdClient.GetValue(ctx, dbKey)
 	if gr != "" {
-		return nil, errors.New(constant.KeyAlreadyPresent)
+		return nil, status.Error(codes.AlreadyExists, errors.New(constant.Etcd+constant.KeyAlreadyPresent).Error())
```
Comment (JaiminRathod) [inline internal/controller/server/repository/metadata/metadata.go:47]: Change as suggested above

File: internal/controller/server/repository/metadata/metadata_test.go [hunk 7]
```diff
@@ -63,7 +66,7 @@ func Test_metadataRepository_Save_KeyAlreadyPresent(t *testing.T) {
 	ctx := context.Background()
 	_, err = testMetadataRepo.Save(ctx, "test data", metadataVal)
 
-	if err.Error() != "key already present" {
+	if err.Error() != status.Error(codes.AlreadyExists, errors.New(constant.Etcd+constant.KeyAlreadyPresent).Error()).Error() {
```
Comment (JaiminRathod) [inline internal/controller/server/repository/metadata/metadata_test.go:69]: Change as suggested above

**Review by JaiminRathod — CHANGES_REQUESTED** [review]

**Review by littlestar642 — APPROVED** [review]

File: internal/controller/server/repository/job/job.go [hunk 8]
```diff
@@ -61,23 +63,26 @@ func (j jobRepository) Save(ctx context.Context, jobID uint64, executionData *cl
...
 // FetchNextJob returns jobID and executionData from pendingList
 func (j jobRepository) FetchNextJob(ctx context.Context) (string, *clientCPproto.RequestForExecute, error) {
 	keys, values, err := j.etcdClient.GetAllKeyAndValues(ctx, constant.JobPendingPrefix)
 	if err != nil {
-		return "", nil, err
+		return "", nil, status.Error(codes.Internal, err.Error())
 	}
 	if len(values) == 0 {
-		return "", nil, errors.New("no pending job in pending job list")
+		return "", nil, status.Error(codes.NotFound, errors.New(constant.Controller+"no pending job in pending job list").Error())
```
Comment (littlestar642) [inline internal/controller/server/repository/job/job.go:79]: i guess in place of errors.New we can just use fmt.Sprintf() or simple string concatenation to join the strings.

File: internal/controller/server/repository/metadata/metadata.go (hunk 6 above)
Comment (littlestar642) [reply internal/controller/server/repository/metadata/metadata.go:47]: use fmt.Sprintf or simple string concatenation

**Review by littlestar642 — CHANGES_REQUESTED** [review]

**Review by JaiminRathod — APPROVED** [review]

**Review by littlestar642 — APPROVED** [review]

**Pull request merged**